src/zkp/pimod.rs

```diff
@@ -371,40 +371,51 @@ fn square_roots_mod_composite(
     n: &BigNumber,
     p: &BigNumber,
     q: &BigNumber,
-) -> Result<[BigNumber; 4]> {
-    let (y1, y2) = square_roots_mod_prime(n, p)?;
-    let (z1, z2) = square_roots_mod_prime(n, q)?;
+) -> Result<Option<[BigNumber; 4]>> {
+    let (y1, y2) = match square_roots_mod_prime(n, p) {
+        Some(roots) => roots,
+        None => return Ok(None),
+    };
+    let (z1, z2) = match square_roots_mod_prime(n, q) {
+        Some(roots) => roots,
+        None => return Ok(None),
+    };
 
     let x1 = chinese_remainder_theorem(&y1, &z1, p, q)?;
     let x2 = chinese_remainder_theorem(&y1, &z2, p, q)?;
     let x3 = chinese_remainder_theorem(&y2, &z1, p, q)?;
     let x4 = chinese_remainder_theorem(&y2, &z2, p, q)?;
 
-    Ok([x1, x2, x3, x4])
+    Ok(Some([x1, x2, x3, x4]))
 }
 
 #[cfg_attr(feature = "flame_it", flame("PaillierBlumModulusProof"))]
 fn fourth_roots_mod_composite(
     n: &BigNumber,
     p: &BigNumber,
     q: &BigNumber,
-) -> Result<Vec<BigNumber>> {
+) -> Result<Option<Vec<BigNumber>>> {
     let mut fourth_roots = vec![];
 
-    let xs = square_roots_mod_composite(n, p, q)?;
-    for x in xs {
-        match square_roots_mod_composite(&x, p, q) {
-            Ok(res) => {
+    let square_roots = match square_roots_mod_composite(n, p, q)? {
+        None => return Ok(None),
+        Some(roots) => roots,
+    };
+
+    // Note: could iter-ize this
+    for root in square_roots {
+        match square_roots_mod_composite(&root, p, q)? {
+            Some(res) => {
                 for y in res {
                     fourth_roots.push(y);
                 }
             }
-            Err(_) => {
+            None => {
                 continue;
             }
         }
     }
```

Comment:
One potential improvement for the future is to make this return type more explicit. There are always either exactly 4 or exactly 0 4th roots returned from this method, so a more specific return type would be `Result<Option<[BigNumber; 4]>>`